src/App.tsx

```diff
@@ -1,223 +1,268 @@
+  const [tempTodoTitle, setTempTodoTitle] = useState('');
+  const [
+    statusToFilter,
+    setStatusToFilter,
+  ] = useState<TodosFilter>(TodosFilter.All);
+
+  const tempTodo = useMemo(() => ({
+    id: 0,
+    userId: 0,
+    title: tempTodoTitle,
+    completed: false,
+  }), [tempTodoTitle]);
+
+  const closeNotification = useCallback(() => setHasError(false), []);
+
+  const filtredTodos = useMemo(() => (
+    todos.filter(({ completed }) => {
+      switch (statusToFilter) {
+        case TodosFilter.Active:
+          return !completed;
+
+        case TodosFilter.Completed:
+          return completed;
+
+        default:
+          return true;
+      }
+    })
+  ), [todos, statusToFilter]);
+
+  const completedTodos = useMemo(() => (
+    todos.filter((todo) => todo.completed)
+  ), [todos]);
+
+  const generateError = useCallback((message: string) => {
+    setErrorMessage(message);
+    setHasError(true);
+  }, []);
+
+  const getTodosFromServer = useCallback(async () => {
+    try {
+      if (user) {
+        const todosFromServer = await getTodos(user.id);
+
+        setTodos(todosFromServer);
+      }
+    } catch (error) {
+      generateError('Unable to show todos');
+    }
+  }, []);
+
+  const addTodoToServer = useCallback(async (todoTitle: string) => {
+    if (user) {
+      try {
+        setIsAdding(true);
+        setTempTodoTitle(todoTitle);
+
+        const addedTodo = await addTodo({
+          title: todoTitle,
+          userId: user.id,
+          completed: false,
+        });
+
+        setTodos(currTodos => [...currTodos, addedTodo]);
+        setIsAdding(false);
+      } catch (error) {
+        generateError('Unable to add todo');
+        setIsAdding(false);
+      }
+    }
+  }, []);
+
+  const removeTodoFromServer = useCallback(async (todoId: number) => {
+    try {
+      setTodoIdsLoading(currIds => [...currIds, todoId]);
+
+      await removeTodo(todoId);
+
+      setTodos(currTodos => (
+        currTodos.filter(({ id }) => id !== todoId)
+      ));
+
+      setTodoIdsLoading(currIds => (
+        currIds.filter((id) => id !== todoId)
+      ));
+    } catch (error) {
+      generateError(`Unable to remove todo ${todoId}`);
+    }
+  }, []);
+
+  const removeAllCompletedTodos = useCallback(async () => {
+    try {
+      await Promise.all(completedTodos.map(({ id }) => (
+        removeTodoFromServer(id)
+      )));
+    } catch (error) {
+      generateError('Unable to remove all completed todo');
+    }
+  }, [completedTodos]);
+
+  const toggleTodoServerStatus = useCallback(async (
+    todoId: number,
+    status: boolean,
+  ) => {
+    try {
+      setTodoIdsLoading(currIds => [...currIds, todoId]);
+
+      await patchTodo(todoId, { completed: status });
+
+      setTodos(currTodos => (
+        currTodos.map(todo => (
+          todo.id === todoId
+            ? ({ ...todo, completed: status })
+            : todo
+        ))
+      ));
```

Comment:
in that case better to use updated todo from server or load new todos than setting this value

I mean in case of validation or any other errors on api you'll use data from server (which should be only source of truth)

Reply:
Okay, will save response from fetch to todos

Reply:
better make a request to server, that is better in case of this application, 'cause we need to have one source of truth, and it's an database on server, additional request is not critical in this case (I was wrong about it on Q&A)

Reply:
I'm a little bit confused, so I can just make one extra request for getting todo after pathTodo?

Reply:
yes, you can do same with other operations with todo (and better to do it)

Sorry for confusing, I was wrong when I tell about avoiding additional requests, in case of this application this solution is better, 'cause we won't face any performance issues if we'll make them, but we'll have an actual information from database

Reply:
It's okay, will try to fix all tomorrow ⭐